Source/JavaScriptCore/wasm/js/JSWebAssemblyStruct.cpp

```diff
@@ -0,0 +1,220 @@
+
+    size_t fieldOffset = 0;
+    for (unsigned i = 0; i < fieldIndex; ++i)
+        fieldOffset += typeKindSize(type->field(i).type.kind);
+    return fieldOffset;
+}
+
+uint64_t JSWebAssemblyStruct::get(uint32_t fieldIndex) const
+{
+    using Wasm::TypeKind;
+
+    ASSERT(fieldIndex < structType()->fieldCount());
+    const uint32_t* targetPointer = reinterpret_cast<const uint32_t*>(m_payload.data() + getFieldOffset(fieldIndex, structType()));
+
+    switch (field(fieldIndex).type.kind) {
+    case TypeKind::I32:
+    case TypeKind::F32:
+        return *targetPointer;
+    case TypeKind::I64:
+        return *reinterpret_cast<const uint64_t*>(targetPointer);
+    case TypeKind::F64:
+        return *reinterpret_cast<const uint64_t*>(targetPointer);
+    case TypeKind::Externref:
+    case TypeKind::Funcref:
+    case TypeKind::Ref:
+    case TypeKind::RefNull:
+        return JSValue::encode(bitwise_cast<WriteBarrierBase<Unknown>*>(targetPointer)->get());
+    default:
+        ASSERT_NOT_REACHED();
+        return 0;
+    }
+}
+
+void JSWebAssemblyStruct::set(JSGlobalObject* globalObject, uint32_t fieldIndex, JSValue argument)
+{
+    using Wasm::TypeKind;
+
+    uint8_t* targetPointer = m_payload.data() + getFieldOffset(fieldIndex, structType());
+    switch (field(fieldIndex).type.kind) {
+    case TypeKind::I32: {
+        *reinterpret_cast<int32_t*>(targetPointer) = argument.toInt32(globalObject);
+        return;
+    }
+    case TypeKind::F32: {
+        *reinterpret_cast<uint32_t*>(targetPointer) = bitwise_cast<uint32_t>(argument.toFloat(globalObject));
+        return;
+    }
+    case TypeKind::I64: {
+        *reinterpret_cast<uint64_t*>(targetPointer) = static_cast<uint64_t>(argument.toBigInt64(globalObject));
+        return;
+    }
+    case TypeKind::F64: {
+        *reinterpret_cast<uint64_t*>(targetPointer) = bitwise_cast<uint64_t>(argument.toNumber(globalObject));
+        return;
+    }
+    case TypeKind::Arrayref:
+    case TypeKind::Externref:
+    case TypeKind::Funcref:
+    case TypeKind::Ref:
+    case TypeKind::RefNull: {
+        bitwise_cast<WriteBarrierBase<Unknown>*>(targetPointer)->set(vm(), this, argument);
+        return;
+    }
+    case TypeKind::Func:
+    case TypeKind::Struct:
+    case TypeKind::Array:
+    case TypeKind::Void:
+    case TypeKind::Rec:
+    case TypeKind::I31ref: {
+        break;
+    }
+    }
+
+    ASSERT_NOT_REACHED();
+}
+
+void JSWebAssemblyStruct::finishCreation(VM& vm)
+{
+    Base::finishCreation(vm);
+    ASSERT(inherits(info()));
+}
+
+
+template<typename Visitor>
+void JSWebAssemblyStruct::visitChildrenImpl(JSCell* cell, Visitor& visitor)
+{
+    Base::visitChildren(cell, visitor);
+
+    auto* wasmStruct = jsCast<JSWebAssemblyStruct*>(cell);
+    for (unsigned i = 0; i < wasmStruct->structType()->fieldCount(); ++i) {
+        if (isRefType(wasmStruct->field(i).type))
+            visitor.append(bitwise_cast<WriteBarrier<Unknown>>(wasmStruct->get(i)));
```

Comment:
This is wrong since it is passing a value, not reference. Need to get WriteBarrier* and pass the reference.